clippy_lints/src/booleans.rs

```diff
@@ -203,6 +203,20 @@ fn check_simplify_not(cx: &LateContext<'_>, msrv: &Msrv, expr: &Expr<'_>) {
         && let Some(suggestion) = simplify_not(cx, msrv, inner)
         && cx.tcx.lint_level_at_node(NONMINIMAL_BOOL, expr.hir_id).0 != Level::Allow
     {
+        use clippy_utils::sugg::{Sugg, has_enclosing_paren};
+        let maybe_par = if let Some(sug) = Sugg::hir_opt(cx, inner) {
+            match sug.maybe_par() {
```

Comment:
So we actually both misunderstood `maybe_par` in our own ways, and my bad. In this case you usually don't match on it, as it essentially adds parentheses around all `MaybeParen` that don't already have them and all `BinOp`s, and turns them into `NonParen` (which now is a `str` with parentheses around it). This is so it can be used in `format!` with a piece of the original code for use in the suggestion.

But that's not the case here, we aren't printing it, only seeing if it needs parentheses. So really, the only thing here that will ever be hit is `NonParen` and `MaybeParen`. Instead, refer to the implementation of [`maybe_par`](https://doc.rust-lang.org/nightly/nightly-rustc/src/clippy_utils/sugg.rs.html#339-355) and copy it a bit. If it's `MaybeParen` (with `has_enclosing_paren == false`) or `BinOp`, then you return true.

Reply:
It seemed like `matches!` couldn't handle both if guards and `|` , so used the usual match.